service/s3/s3manager/batch.go

```diff
@@ -288,18 +297,64 @@ func (iter *DeleteObjectsIterator) DeleteObject() BatchDeleteObject {
 // Once the batch size is met, this will call the deleteBatch function.
 func (d *BatchDelete) Delete(ctx aws.Context, iter BatchDeleteIterator) error {
 	var errs []Error
+	objects := []BatchDeleteObject{}
+	var input *s3.DeleteObjectsInput
+
 	for iter.Next() {
-		object := iter.DeleteObject()
-		if _, err := d.Client.DeleteObjectWithContext(ctx, object.Object); err != nil {
-			errs = append(errs, newError(err, object.Object.Bucket, object.Object.Key))
+		o := iter.DeleteObject()
+
+		if input == nil {
+			input = initDeleteObjectsInput(o.Object)
 		}
 
-		if object.After == nil {
-			continue
+		parity := hasParity(input, o)
+		if parity {
+			input.Delete.Objects = append(input.Delete.Objects, &s3.ObjectIdentifier{
+				Key:       o.Object.Key,
+				VersionId: o.Object.VersionId,
+			})
+			objects = append(objects, o)
 		}
 
-		if err := object.After(); err != nil {
-			errs = append(errs, newError(err, object.Object.Bucket, object.Object.Key))
+		if len(input.Delete.Objects) == d.BatchSize || !parity {
+			if err := deleteBatch(d, input); err != nil {
+				errs = append(errs, err...)
+			}
+
+			for _, object := range objects {
```

Comment:
It would be nice for this for loop to be included in the `deleteBatch` function, or as its own function so the logic isn't duplicated in here.